File: pyorion/api/controlcenter.py
# Copyright 2025-2030 Ari Bermeki @ YellowSiC within The Commons Conservancy
# SPDX-License-Identifier: Apache-2.0
# SPDX-License-Identifier: MIT

"""ControlCenter API - System notifications and related operations.

Wrapper around the Rust event loop backend to send notifications
and interact with the system Control Center.
"""

from typing import Any, Optional

from pyorion.runtime.runtime_handle import event_register


class ControlCenterAPI:
    """Asynchronous API wrapper for system Control Center operations."""

    def __init__(self) -> None:
        """Initialize a new :class:`ControlCenterAPI` instance."""

    async def notification(
        self,
        summary: str,
        body: Optional[str] = None,
        app_id: Optional[str] = None,
        appname: Optional[str] = None,
        icon: Optional[str] = None,
        auto_icon: Optional[bool] = None,
        image_path: Optional[str] = None,
        sound_name: Optional[str] = None,
        subtitle: Optional[str] = None,
        timeout: Optional[int] = None,
        notification_id: Optional[int] = None,
        action: Optional[tuple[str, str]] = None,
    ) -> Optional[dict[str, Any]]:
        """Send a system notification through the Control Center.

        :param summary: The notification summary (title text).
        :type summary: str
        :param body: The body text of the notification.
        :type body: Optional[str]
        :param app_id: Optional application identifier.
        :type app_id: Optional[str]
        :param appname: Application name to display in the notification.
        :type appname: Optional[str]
        :param icon: Path to an icon file.
        :type icon: Optional[str]
        :param auto_icon: Whether to automatically choose an icon.
        :type auto_icon: Optional[bool]
        :param image_path: Path to an image to include in the notification.
        :type image_path: Optional[str]
        :param sound_name: Name of the sound to play with the notification.
        :type sound_name: Optional[str]
        :param subtitle: Subtitle or secondary text.
        :type subtitle: Optional[str]
        :param timeout: Timeout in milliseconds before the notification disappears.
        :type timeout: Optional[int]
        :param notification_id: ID of the notification to replace, if any.
        :type notification_id: Optional[int]
        :param action: Tuple of (action_key, label) for notification action button.
        :type action: Optional[tuple[str, str]]

        :return: Notification metadata as a dictionary, or ``None`` if unsupported.
        :rtype: Optional[dict[str, Any]]
        """
        args = [
            summary,
            body,
            app_id,
            appname,
            icon,
            auto_icon,
            image_path,
            sound_name,
            subtitle,
            timeout,
            notification_id,
            action,
        ]
        return await event_register("controlcenter.notification", args=args)

File: src/api/control_center.rs
// Copyright 2025-2030 PyFrame Programme within The Commons Conservancy
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

use crate::api_manager::ApiManager;
use anyhow::Result;
use notify_rust::Notification;
use pyorion_macros::api;

pub fn control_center_api(_api_manager: &mut ApiManager) {
    _api_manager.register_api("controlcenter.notification", notification);
}

#[api]
fn notification(
    summary: String,
    body: Option<String>,
    app_id: Option<String>,
    appname: Option<String>,
    icon: Option<String>,
    auto_icon: Option<bool>,
    image_path: Option<String>,
    sound_name: Option<String>,
    subtitle: Option<String>,
    timeout: Option<i32>,
    id: Option<u32>,
    action: Option<(String, String)>,
) -> Result<()> {
    let mut binding = Notification::new();
    let notify = binding.summary(&summary);

    if let Some(body) = body {
        notify.body(&body);
    }

    #[cfg(target_os = "windows")]
    if let Some(app_id) = app_id {
        notify.app_id(&app_id);
    }

    #[cfg(not(target_os = "windows"))]
    let _ = app_id;

    if let Some(appname) = appname {
        notify.appname(&appname);
    }

    #[cfg(target_os = "windows")]
    if let Some(path) = image_path {
        notify.image_path(&path);
    }

    #[cfg(not(target_os = "windows"))]
    let _ = image_path;

    #[cfg(any(target_os = "windows", target_os = "macos"))]
    if let Some(sound) = sound_name {
        notify.sound_name(&sound);
    }

    #[cfg(not(any(target_os = "windows", target_os = "macos")))]
    let _ = sound_name;

    if let Some(subtitle) = subtitle {
        notify.subtitle(&subtitle);
    }

    if let Some(id) = &id {
        notify.id(*id);
    }

    if let Some(true) = auto_icon {
        notify.auto_icon();
    }

    if let Some(icon) = icon {
        notify.icon(&icon);
    }

    if let Some(timeout) = timeout {
        notify.timeout(timeout);
    }

    if let Some((identifier, label)) = &action {
        notify.action(identifier, label);
    }

    notify.show()?;
    Ok(())
}
